include/rdma/fi_profile.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2023 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_PROFILE_H
+#define FI_PROFILE_H
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include <rdma/fi_domain.h>
+
+/*
+ * pre-defined profiling variables
+ */
+enum {
+	FI_VAR_UNEXP_MSG_CNT,    // type: uint64_t
+};
+
+/*
+ * pre-defined events which can change the profiling variables
+ */
+enum {
+	FI_EVENT_UNEXP_MSG_RECVD,    // var = FI_VAR_UNEXP_MSG_CNT
+	FI_EVENT_UNEXP_MSG_MATCHED,  // var = FI_VAR_UNEXP_MSG_CNT
+};
+
+struct fi_profile_desc {
+	uint32_t id;
+	enum fi_datatype type;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+
+struct fid_profile;
+
+struct fi_profile_ops {
+	size_t size;
+	void (*reset)(struct fid_profile *prof_fid,  uint64_t flags);
+	int (*query_vars)(struct fid_profile *prof_fid,
+	                     struct fi_profile_desc *varlist, int *count);
+	int (*query_events)(struct fid_profile *prof_fid,
+	                     struct fi_profile_desc *eventlist, int *count);
+	int (*read_var)(struct fid_profile *prof_fid, uint32_t var_id, void *data);
+	int (*reg_callback)(struct fid_profile *prof_fid, uint32_t event,
+	            int (*callback)(struct fid_profile *prof_fid, uint32_t event,
+	                    void *param, size_t size, void *context), 
+	            void *context);
+	void (*lock)(struct fid_profile *prof_fid);
+	void (*unlock)(struct fid_profile *prof_fid);
```

Comment:
I was thinking about this last night, and I would rename these functions to start_reads / end_reads, plus add a flags field.  This more clearly indicates what the app is trying to do, without implying any specific implementation on the part of the provider.

Reply:
Ok, will change lock/unlock to start_reads/end_reads.